src/Dapr.AspNetCore/DaprAuthenticationHandler.cs

```diff
@@ -0,0 +1,63 @@
+﻿// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+namespace Dapr.AspNetCore
+{
+    using System.Collections.Generic;
+    using System.Security.Claims;
+    using System.Text.Encodings.Web;
+    using System.Threading.Tasks;
+    using Microsoft.AspNetCore.Authentication;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Extensions.Options;
+
+    internal class DaprAuthenticationHandler : AuthenticationHandler<DaprAuthenticationHandlerOptions>
+    {
+        const string DaprApiToken = "Dapr-Api-Token";
+
+        public DaprAuthenticationHandler(
+            IOptionsMonitor<DaprAuthenticationHandlerOptions> options,
+            ILoggerFactory logger,
+            UrlEncoder encoder, 
+            ISystemClock clock) : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            return Task.FromResult(HandleAuthenticate());
+        }
+
+        private AuthenticateResult HandleAuthenticate()
+        {
+            if (!Request.Headers.TryGetValue(DaprApiToken, out var token))
+            {
+                return AuthenticateResult.NoResult();
+            }
+
+            var expectedToken = Options.TokenFactory?.Invoke();
```

Comment:
This is an interesting behavioral question - hopefully @Tratcher can answer as well.

So if I configure the new auth handler, but fail to set the expected today via `DAPR_API_TOKEN` it looks like this would always return `NoResult()`. So does that mean that the request is considered unauthenticated?

Reply:
Elsewhere I recommended moving the env var resolution to Options and setting it from IConfiguration at startup. That would also be the place to add validation for a missing value. E.g. crash on startup for bad config.

Reply:
Token is now a string property on `Options`. I changed to `Fail` with a error message.